pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+	nodesSynced cache.InformerSynced
+	// clusterCIDRLister is able to list/get clustercidrs and is populated by the shared informer passed to controller.
+	clusterCIDRLister networkinglisters.ClusterCIDRLister
+	// clusterCIDRSynced returns true if the clustercidr shared informer has been synced at least once.
+	clusterCIDRSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
+	recorder              record.EventRecorder
+	// queue is where incoming work is placed to de-dup and to allow "easy"
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+
+	// lock guards nodesInProcessing and cidrMap to avoid races in CIDR allocation.
+	lock *sync.Mutex
+	// nodesInProcessing is a set of nodes that are currently being processed.
+	nodesInProcessing sets.String
+	// cidrMap maps ClusterCIDR labels to internal ClusterCIDR objects.
+	cidrMap map[string][]*cidrset.ClusterCIDR
+}
+
+// NewMultiCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one for each ip family).
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	clusterCIDRInformer networkinginformers.ClusterCIDRInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList,
+	testCIDRMap map[string][]*cidrset.ClusterCIDR,
+) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("client is nil")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventSource := v1.EventSource{
+		Component: "multiCIDRRangeAllocator",
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, eventSource)
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Started sending events to API Server. (EventSource = %v)", eventSource)
+
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                client,
+		nodeLister:            nodeInformer.Lister(),
+		nodesSynced:           nodeInformer.Informer().HasSynced,
+		clusterCIDRLister:     clusterCIDRInformer.Lister(),
+		clusterCIDRSynced:     clusterCIDRInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                  &sync.Mutex{},
```

Comment:
nit, you don't need initialize mutex

https://golang.org/doc/effective_go.html#data